### PR TITLE
ARROW-16011: [R] CI jobs should fail if lintr picked up issues 

### PR DESCRIPTION
**Pull request by assignUser**
Building on @thisisnic 's work  #12696 
This requires less maintenance in the workflows but does add a `Suggests` on lintr

### CODE REVIEW

**Comment by assignUser** [comment]
@thisisnic could you take a look at this?

File: .github/workflows/r.yml [hunk 1]
```diff
@@ -248,6 +248,7 @@ jobs:
     env:
       ARROW_R_CXXFLAGS: "-Werror"
       _R_CHECK_TESTS_NLINES_: 0
+      GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN}}
```
Comment (thisisnic) [inline .github/workflows/r.yml:251]: Now you're using a different method to the one I started, is this line still needed?

File: r/tests/testthat/test-style.R [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+test_that("Package is lint free", {
+    skip_on_cran()
+    skip_if_not_installed("lintr")
+    lintr::expect_lint_free()
```
Comment (thisisnic) [inline r/tests/testthat/test-style.R:21]: Nice, I didn't know this function existed before now!

File: r/R/array-data.R [hunk 3]
```diff
@@ -44,7 +44,7 @@
 ArrayData <- R6Class("ArrayData",
   inherit = ArrowObject,
   active = list(
-    type = function() ArrayData__get_type(self),
+    type =function() ArrayData__get_type(self),
```
Comment (thisisnic) [inline r/R/array-data.R:47]: Just adding a comment here to remind us to remove this once the PR is ready to be approved

File: .github/workflows/r.yml [hunk 4]
```diff
@@ -144,7 +144,7 @@ jobs:
           # Don't set a TZ here to test that case. These builds will have the following warning in them:
           #   System has not been booted with systemd as init system (PID 1). Can't operate.
           #   Failed to connect to bus: Host is down
-          archery docker run -e TZ="" r
+          archery docker run -e TZ="" -e GITHUB_ACTIONS="true" r
```
Comment (thisisnic) [inline .github/workflows/r.yml:147]: Please can you just remind me what this does?

**Review by thisisnic — COMMENTED** [review]
A few questions/comments, but I like the use of `lintr::expect_lint_free()`.  The only question I have about that is - given that when I run `lintr::lint_package()` it takes a while to run, are there concerns about this taking longer to run our tests or is it not worth worrying about?

File: .github/workflows/r.yml (hunk 4 above)
Comment (assignUser) [reply .github/workflows/r.yml:147]: Without this `lintr` does not detect being on GHA and no annotation will be generated.

**Comment by assignUser** [comment]
This should have failed but `lintr` could not find the package, need to set the working directory correctly.

**Comment by assignUser** [comment]
@thisisnic could you approve the checks please :)

**Comment by thisisnic** [comment]
Thanks for doing this @assignUser, this is really helpful!  Looks good now, I'll remove the intentional CI failure and then merge once the CI passes.

File: r/R/array-data.R (hunk 3 above)
Comment (thisisnic) [inline r/R/array-data.R:47]:
```suggestion
    type = function() ArrayData__get_type(self),
```

**Review by thisisnic — APPROVED** [review]

**Pull request closed without merging**